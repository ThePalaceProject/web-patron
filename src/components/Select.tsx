// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import * as React from "react";
import SvgExpandMore from "icons/ExpandMore";

type SelectProps = React.ComponentProps<"select">;
const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
  ({ children, ...props }, ref) => {
    return (
      <div sx={{ position: "relative" }}>
        <select
          ref={ref}
          sx={{
            appearance: "none",
            width: "100%",
            pl: 2,
            pr: "2rem",
            py: 1,
            border: "solid",
            borderRadius: 1,
            variant: "text.body.regular",
            fontSize: "-1"
          }}
          {...props}
        >
          {children}
        </select>
        <SvgExpandMore
          sx={{
            position: "absolute",
            height: "100%",
            width: "1.75rem",
            right: "0.25rem",
            pointerEvents: "none"
          }}
        />
      </div>
    );
  }
);
Select.displayName = "Select";

export const Label: React.FC<React.ComponentProps<"label">> = ({
  children,
  ...props
}) => {
  return (
    <label sx={{ whiteSpace: "nowrap" }} {...props}>
      {children}
    </label>
  );
};

export default Select;
