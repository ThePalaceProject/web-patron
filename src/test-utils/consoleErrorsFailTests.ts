import { jest } from "@jest/globals";
const consoleErrorSpy = jest.spyOn(global.console, "error");

// use this to make console errors fail tests. Useful when running in to issues.
export default function consoleErrorsFailTests() {
  consoleErrorSpy.mockImplementation((msg, ...opts) => {
    console.warn(msg, ...opts);
    throw new Error(msg);
  });
}
