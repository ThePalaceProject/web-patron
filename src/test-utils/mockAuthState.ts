import { AuthCredentials, OPDS1 } from "interfaces";
import Cookie from "js-cookie";
import { generateCredentials } from "utils/auth";
import { basicTokenAuthenticationUrl } from "./fixtures";

const mockCookie = Cookie as any;

export const creds: AuthCredentials = {
  token: "some-token",
  methodType: OPDS1.BasicAuthType
};

export const persistentUserCredentials = generateCredentials("1234", "pinpin");
export const firstToken = "IaMaBeArErToKeN";
export const tokenCreds1: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: firstToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

export const newToken = "IaMaBeArErToKeN2";
export const tokenCreds2: AuthCredentials = {
  token: {
    basicToken: persistentUserCredentials,
    bearerToken: newToken
  },
  authenticationUrl: basicTokenAuthenticationUrl,
  methodType: OPDS1.BasicTokenAuthType
};

const str = JSON.stringify;

export default function mockAuthenticatedOnce(
  credentials: AuthCredentials | null = creds
) {
  mockCookie.get.mockReturnValueOnce(str(credentials));
}

export function mockAuthenticated(credentials: AuthCredentials | null = creds) {
  mockCookie.get.mockReturnValue(str(credentials));
}
