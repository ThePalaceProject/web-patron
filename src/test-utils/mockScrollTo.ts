import { jest } from "@jest/globals";
/**
 * JSdom doesn't include a stub for the scrollTo function, which we use inside of
 * Lane. Therefore we will stub it globally here
 */

Element.prototype.scrollTo = jest.fn();
window.scrollTo = jest.fn();

export default {};
