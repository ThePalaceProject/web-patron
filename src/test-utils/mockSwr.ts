import { jest } from "@jest/globals";
import { responseInterface } from "swr";

export function makeSwrResponse<T>(value?: Partial<responseInterface<T, any>>) {
  return {
    data: undefined,
    error: undefined,
    revalidate: jest.fn(),
    isValidating: false,
    mutate: jest.fn(),
    ...value
  };
}

export type MockSwr<T> = (value?: Partial<responseInterface<T, any>>) => void;
