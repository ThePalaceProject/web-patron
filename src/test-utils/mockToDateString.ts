import { jest } from "@jest/globals";
/**
 * This file mocks Date.prototype.toDateString so that it just returns a
 * stub string. This makes code using that function testable across timezones
 * since we can't reliably set the timezone of a node process to something specific.
 */
export const MOCK_DATE_STRING = "mock-date-string";
export const toDateStringSpy = jest
  .spyOn(Date.prototype, "toDateString")
  .mockImplementation(() => MOCK_DATE_STRING);
