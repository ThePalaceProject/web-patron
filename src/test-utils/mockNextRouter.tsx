// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import { jest } from "@jest/globals";
import * as React from "react";
import Router from "next/router";
import { RouterContext } from "next/dist/shared/lib/router-context.shared-runtime";
import { NextRouter } from "next/router";
import { libraryData } from "test-utils/fixtures";

/**
 * Mock for the next/Router import.
 */

export const mockPush = jest.fn().mockImplementation(async () => true);
Router.push = mockPush;

export const MockNextRouterContextProvider: React.FC<{
  router?: Partial<NextRouter>;
}> = ({ router = {}, children }) => {
  const {
    isReady = true,
    basePath = "",
    route = "",
    pathname = "",
    query = {
      // add the default library slug to the default query params
      library: libraryData.slug
    },
    // default as path is the home page
    asPath = `/${libraryData.slug}`,
    push = mockPush,
    replace = jest.fn().mockImplementation(async () => true),
    reload = jest.fn().mockImplementation(() => null),
    back = jest.fn().mockImplementation(() => null),
    prefetch = jest.fn().mockImplementation(async () => undefined),
    beforePopState = jest.fn().mockImplementation(() => null),
    isFallback = false,
    isLocaleDomain = false,
    isPreview = false,
    events = {
      on: () => null,
      off: () => null,
      emit: () => null
    }
  } = router;
  return (
    <RouterContext.Provider
      value={{
        isReady,
        basePath,
        route,
        pathname,
        query,
        asPath,
        push,
        replace,
        reload,
        back,
        prefetch,
        beforePopState,
        isFallback,
        isLocaleDomain,
        isPreview,
        events
      }}
    >
      {children}
    </RouterContext.Provider>
  );
};

export default Router;
