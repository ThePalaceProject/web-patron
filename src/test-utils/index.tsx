import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { beforeEach, expect, jest } from "@jest/globals";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import serializer from "jest-emotion";
import { Provider as ReakitProvider } from "reakit";
import { MockNextRouterContextProvider } from "./mockNextRouter";
import { NextRouter } from "next/router";
import { enableFetchMocks } from "jest-fetch-mock";
import { LibraryProvider } from "components/context/LibraryContext";
import { BreadcrumbProvider } from "components/context/BreadcrumbContext";
import { UserContext, UserState } from "components/context/UserContext";
import { ThemeProvider } from "theme-ui";
import makeTheme from "theme";
import mockConfig from "test-utils/mockConfig";
import track from "analytics/track";
import "react-intersection-observer/test-utils";
import "test-utils/mockToDateString";

enableFetchMocks();
expect.addSnapshotSerializer(serializer);

// standard config mock
beforeEach(() => {
  mockConfig();
});

// mock the error tracker to prevent unnecessary console logs
export const mockTrackError = jest.fn();
track.error = mockTrackError;

export { fixtures };

// configure the enzyme adapter
configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
(global as any)["URL"].createObjectURL = mockCreateObjectURL;

type CustomRenderOptions = Parameters<typeof render>[1] & {
  router?: Partial<NextRouter>;
  library?: Partial<LibraryData>;
  user?: Partial<UserState>;
};
/**
 * Our custom render function wraps components in mocked versions of our
 * providers
 */
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const library: LibraryData = {
    ...fixtures.libraryData,
    ...options?.library
  };
  const theme = makeTheme(library.colors);

  const user: UserState = {
    ...fixtures.user,
    ...options?.user
  };

  const AllTheProviders: React.FC = ({ children }) => {
    return (
      <MockNextRouterContextProvider router={options?.router}>
        <ThemeProvider theme={theme}>
          <ReakitProvider>
            <LibraryProvider library={library}>
              <UserContext.Provider value={user}>
                <BreadcrumbProvider>{children}</BreadcrumbProvider>
              </UserContext.Provider>
            </LibraryProvider>
          </ReakitProvider>
        </ThemeProvider>
      </MockNextRouterContextProvider>
    );
  };
  return {
    ...render(ui, { wrapper: AllTheProviders, ...options })
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
