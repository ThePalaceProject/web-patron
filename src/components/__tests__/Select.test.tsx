import * as React from "react";
import { expect, test } from "@jest/globals";
import { render } from "../../test-utils";
import Select, { Label } from "../Select";

test("select matches snapshot", () => {
  expect(
    render(
      <Select>
        <option>op 1</option>
      </Select>
    ).container.firstChild
  ).toMatchSnapshot();
});

test("select label matches snapshot", () => {
  expect(render(<Label>hi from label</Label>).container.firstChild).
toMatchInlineSnapshot(`
.emotion-0 {
  white-space: nowrap;
}

<label
  class="emotion-0"
>
  hi from label
</label>
`);
});
