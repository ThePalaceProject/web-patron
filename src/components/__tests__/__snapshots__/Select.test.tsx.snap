// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select matches snapshot 1`] = `
.emotion-2 {
  position: relative;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  width: 100%;
  padding-left: 8px;
  padding-right: 2rem;
  padding-top: 4px;
  padding-bottom: 4px;
  border: 1px solid #bdbdbd;
  border-radius: 2px;
  font-family: 'Open Sans',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  font-weight: 300;
  line-height: 1.5;
}

.emotion-1 {
  position: absolute;
  height: 100%;
  width: 1.75rem;
  right: 0.25rem;
  pointer-events: none;
}

<div
  class="emotion-2"
>
  <select
    class="emotion-0"
  >
    <option>
      op 1
    </option>
  </select>
  <svg
    class="emotion-1"
    fill="currentColor"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M24 24H0V0h24v24z"
      fill="none"
      opacity="0.87"
    />
    <path
      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6-1.41-1.41z"
    />
  </svg>
</div>
`;
